### ISSUE TITLE
Retain cycles

### ISSUE THREAD

**Issue by npvisual** [DeclarativeHub/Bond#558]
### Context
Currently using a collection view to present a list of item categories. When a category cell is selected (tapped) a detailed view controller is pushed onto the navigation stack to present the list of to-dos in that category.

Similar to the example given in #557 , each to-do cell contains a button to complete the to-do. If you tap that cell's completion button, the item is marked as completed on the backend and the button's image changes to show a checkmark.

The `todoItem` has a special field, `completed`, which is a signal of Bool, to catch any changes to the state of the `todoItem`. This is then bound to the button's `isSelected` reactive extension.

All of this works well within the selected category.

```
todoItems.bind(to: collectionView, cellType: TodoItemCell.self) { (cell, todoItem) in
    // Dispose the cell's bag to dispose any previous bindings made to the cell
    cell.reactive.bag.dispose()
    
    cell.titleLabel.text = todoItem.name
    todoItem.completed
        .toSignal()
        .bind(to: cell.completeTodoItemButton.reactive.isSelected)
        .dispose(in: cell.reactive.bag)

    cell.completeTodoItemButton.reactive.tap
        .with(latestFrom: category.selected)
        .flatMapLatest { _, catg in 
            todoService.complete(todoItem, catg) }
        .bind(to: cell) { cell in 
            print("Completed task \(todoItem)")
        }
        .dispose(in: cell.reactive.bag) // Put the disposable into the bag
}
```

### Problem
Now suppose you've completed to-dos # 3 and # 4 (i.e tapped the buttons for cells 3 and 4 in that category).  And then you use the back button (part of the standard navigation controller flow) to come back to the category list and select a different category.

As soon as you select that new category, item # 3 and 4 of that specific category get updated in the backend as "completed" -- even though **NO** buttons were pressed for this category.

When I use `debug()` on all the different signals, it appears that none of them ever get disposed when the back button is pressed (to get back to the list of categories) -- which I would expect is necessary to avoid those retain cycles.

Even if I force a call to `collectionView.dispose()` when the back button is pressed, nothing gets disposed.

**And** considering that the update to the backend DB can only happen from within the `cell.completeTodoItemButton.reactive.tap` block, which should only be triggered if the `completeTodoItemButton` is pressed, I am a little baffled.

Any ideas about what's going on ?

**Comment by npvisual** [comment]
One interesting piece of information : even if I add a breakpoint at `todoService.complete` (in the flatMap), this second "shadow" call isn't caught -- however the call is stopped by the debugger during the `tap` in the first category).

**Comment by srdanrasic** [comment]
Binding within a cell should be disposed when the cell is deallocated which should happen when the parent collection view is deallocated. The collection view should be deallocated when the view controller is deallocated which should should happen when it's dismissed/popped. Try adding `deinit` to the view controller and see if it gets called. 

Something is probably holding the view controller or a collection view from getting deallocated. I don't see anything wrong in the code you posted, so it's probably elsewhere. 

It super weird behaviour you are experiencing. Are you sharing a service instance between categories?

**Comment by frankois944** [comment]
Hey, 
thinks to add unowned / weak self when using closure.
https://medium.com/@abhimuralidharan/functional-programming-closure-reference-cycle-and-fix-f42cc53c6213
Hope it will help

**Comment by npvisual** [comment]
Thanks for the pointers @frankois944. Good read. Since I am mostly using closures for `bind()` and some other ReactiveKit / Bond functions, I went over my code several times, but that didn't really help.

However, that got me looking into the Xcode memory graph and I was definitely able to see the retained view controllers and cells growing over and over as I selected more categories.

So... I went back to the drawing board and went through a process of elimination.

### Solution

Well, it was (of course) pretty darn simple... For some weird reason (probably cut & paste) I had failed to inject the proper reference in some of my bind statements. So instead of using the parameter local to the closure, I was using the outer parameter. For example : 

```
        session.someService.isConditionTrue.bind(to: controller) { ctrl, value in
                controller.removeToDo.isHidden = !value
        }

```

changing that to : 

```
        session.someService.isConditionTrue.bind(to: controller) { ctrl, value in
                ctrl.removeToDo.isHidden = !value
        }

```

(and in a few other places) did the trick.

Thanks @srdanrasic  and @frankois944 for the help.
